lib/pavilion/string_parser.py

```diff
@@ -0,0 +1,460 @@
+#  OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY
+#  OF SUCH DAMAGE.
+#
+#  ###################################################################
+
+# The module file contains the functions and classes needed for parsing
+# strings and performing variable insertion in pavilion configurations.
+#
+# While we don't use a context free grammer for parsing, it is convenient for
+# describing the format:
+
+# # pav strings can contain variable references, escapes, or sub strings.
+# PAV_STR  -> TEXT | TEXT VAR PAV_STR | TEXT ESC PAV_STR | TEXT SUB_STR PAV_STR
+# # Text is anything that doesn't start an escape, sub string, or variable reference
+# # It may also be an empty string.
+# TEXT     -> [^\{[]*
+# # Variable references are a bracket surrounded variable name. These can consist
+# # of a var_set name, variable name, variable index, and sub-variable name. All but the
+# # variable name are optional.
+# VAR      -> {VAR_KEY}
+# VAR_KEY  -> VAR_BASE | VAR_BASE.VAR_IDX | VAR_BASE.VAR_ID | VAR_BASE.VAR_IDX.VAR_ID
+# VAR_BASE -> VAR_ID | VAR_ID.VAR_ID
+# VAR_ID   -> [a-z][a-z0-9_-]+
+# VAR_IDX  -> [0-9]+
+# #
+# ESCAPE   -> \.
+# # A substring is a pav string surrounded by squared brackets. It's used to
+# # denote an area that can be copied as a unit when used with multi-valued variables.
+# SUB_STR  -> [PAV_STR] | [PAV_STR:SEP]
+# SEP      -> .
+
+import re
+import variables
+import config_format
+
+
+class ScanError(ValueError):
+    """Error scanning and tokenizing a Pav string."""
+    def __init__(self, message, error_start, error_end):
+        self.message = message
+        self.error_start = error_start
+        self.error_end = error_end
+
+    def __str__(self):
+        return self.message
+
+
+class ParseError(ScanError):
+    """Error parsing the Pav string tokens"""
+
+
+class ResolveError(ValueError):
+    """Error resolving string variables."""
+
+
+def parse(string):
+    tokens = tokenize(string)
+
+    return PavString(tokens)
+
+
+TEXT_END_RE = re.compile(r'\[|\\|{|}|(:.)?\]')
+
+
+def tokenize(string):
+    tokens = [TextToken('', 0, 0)]
+
+    pos = 0
+
+    while pos < len(string):
+
+        # Grab everything that isn't a special char.
+        match = TEXT_END_RE.search(string, pos)
+        if match is None:
+            # The text went to the end of the string.
+            tokens.append(TextToken(string[pos:], pos, len(string)))
+            break
+        else:
+            # Grab the text we found, even if it's empty.
+            end = match.start()
+            if end != pos:
+                tokens.append(TextToken(string[pos:end], pos, end))
+            pos = end
+
+        end_str = match.group()
+
+        # Process the non-text token (in most cases) that follows.
+        if end_str == '[':
+            tokens.append(SubStringStartToken(pos, pos+1))
+            pos += 1
+        elif end_str.endswith(']'):
+
+            if len(end_str) == 3:
+                separator = end_str[1]
```

Comment:
I may be severely lost, but is the precluding the possibility of a multi-character separator?